### PR TITLE
Implement caching for download of virtual machine images

### PR DESCRIPTION
**Pull request by dilyar85**
This patch introduces an in-memory cache to store and manage downloaded OVF for virtual machine images. It also cleans up stale cached items after a specified time (currently set to 30 minutes). This avoids unnecessary requests to vCenter when the same content library is associated with multiple namespaces within a cluster and speeds up vm image reconciliation time.

Testing Done:
- Internal kind-based e2e tests (both fast and full) passed
- Manually deployed this patch in a WCP testbed (with Image-Registry FSS enabled) and verified the expected workflow by associating a content library to multiple namespaces simultaneously. Logs from VM-Operator (in verbose mode):
```bash
I0207 09:15:21.266601       1 vmprovider.go:248] vsphere "msg"="Cache item miss, downloading OVF from vCenter" "contentVersion"="2" "itemID"="732f3e46-4699-4b66-aa6c-2740b0bfe6ed"
...
I0207 09:15:29.134953       1 vmprovider.go:264] vsphere "msg"="Cache item put" "contentVersion"="2" "itemID"="732f3e46-4699-4b66-aa6c-2740b0bfe6ed" "putResult"=1
I0207 09:15:29.135280       1 vmprovider.go:246] vsphere "msg"="Cache item hit, using cached OVF" "contentVersion"="2" "itemID"="732f3e46-4699-4b66-aa6c-2740b0bfe6ed"
I0207 09:15:29.136055       1 vmprovider.go:246] vsphere "msg"="Cache item hit, using cached OVF" "contentVersion"="2" "itemID"="732f3e46-4699-4b66-aa6c-2740b0bfe6ed"
```

### CODE REVIEW

**Comment by akutz** [comment]
Instead of making the cache part of the provider, you could create a context key for the cache and store/retrieve the cache via the context. It would enable better testing as well as remove the need to pass around the cache everywhere outside of access to the provider struct.

image

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_utils.go
Comment (zyiyi11) [inline pkg/vmprovider/providers/vsphere/contentlibrary/content_library_utils.go:115]: what is the point of switching cases when setting it to the same value? Are you trying to guard against other types than vmi/cvmi?

**Review by zyiyi11 — COMMENTED** [review]
LGTM

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_utils.go
Comment (akutz) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_utils.go:115]: Because he's casting it, and if you combine them into a single case, it won't let you access the fields of that type.

Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_utils.go:115]: What Andrew said :) Btw, this is from the existing code and I am just moving it to a separate function in the `content_library_utils.go` file (as it doesn't contain any dependency on the client).

**Comment by dilyar85** [comment]
> Instead of making the cache part of the provider, you could create a context key for the cache and store/retrieve the cache via the context. It would enable better testing as well as remove the need to pass around the cache everywhere outside of access to the provider struct.

@akutz As previously discussed offline, I am going to keep the current implementation since it doesn't introduce any global package variables. Could you please take another look when you get a chance? Thank you.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (ericvmw) [inline pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:19]: Consider setting a max number of items that can be cached, potentially the cache might take some good memory space. If there are 1000 items with 100kb OVF descriptor in the cluster, then the cache may take 100mb if they may be cached around the same time period.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:54]: Added a separate block to avoid referencing it when `item == nil`.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:19]: That's a fair point. I applied a max size and set it to 100 in the new patch.

File: pkg/vmprovider/providers/vsphere/vmprovider.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/vmprovider.go:208]: Thanks for pointing this out. I updated it to store the content library version directly from fetched item in the new patch.

**Review by bryanv — COMMENTED** [review]
How effective is this cache on pod restart? Is this mostly effective when creating a new namespace with CL already for other namespaces? When there are more items than `maxSize` and during the re-sync, I'm not sure if updating `lastAccessTime` is really cache friend to how we'll access the cache.

Is there any practical benefit of caching on behalf of cluster scoped images? Sharing a CL between cluster and namespace scoped doesn't seem like a likely scenario, if not disallowed in VC.

Ultimately I think we need a friendlier API from CL, especially for when a CL is shared among multiple clusters.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (ericvmw) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:20]: CL validates OVF descriptor to ensure less than 4mb per OVF spec, most OVF template size should be way less than that. Tracking the total size of all cache entries can be also helpful, so it can have the cache bounded by some fixed or configurable memory space as well.

File: pkg/vmprovider/providers/vsphere/vmprovider.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/vmprovider.go:211]: Fair point. Changed it to `if item == nil || item.Type != library.ItemTypeOVF` with the error message updated.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:36]: This is just for removing items that have been cached for a while (longer than the itemExpirationSec) so that we can free some space in the cache. Otherwise, there is no way to add more items after it reaches max size.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:20]: In the worst case then, the cache can grow up to 4MB * 100 = 400MB which is around 381MiB, less than the current 500MiB limit on the container. I checked the TKG CL and it shows as 4.23 MB in total. In case we encounter any OOM from this cache, we can configure a smaller max size and shorter expiration time to clean up items more frequently.

**Comment by akutz** [comment]
> How effective is this cache on pod restart? Is this mostly effective when creating a new namespace with CL already for other namespaces? When there are more items than `maxSize` and during the re-sync, I'm not sure if updating `lastAccessTime` is really cache friend to how we'll access the cache.
> 
> Is there any practical benefit of caching on behalf of cluster scoped images? Sharing a CL between cluster and namespace scoped doesn't seem like a likely scenario, if not disallowed in VC.
> 
> Ultimately I think we need a friendlier API from CL, especially for when a CL is shared among multiple clusters.

@dilyar85 , Please respond to @bryanv here. I know we've discussed answers to some of his questions.

File: controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go
Comment (akutz) [inline controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go:263]: The moment we've added the 3rd param we should decide if an options struct is worth it. For example:

```go
type SyncVirtualMachineImageOpts struct {
  ContentLibraryItemID string
  LatestVersion        string
  VirtualMachineImage  client.Object
}
```

Something like that. And to be clear, the `context` should never go in an options struct.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (akutz) [inline pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:1]: Big comment incoming on this.

**Review by akutz — CHANGES_REQUESTED** [review]

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (akutz) [inline pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:37]: Don't log here. Having a logger in a goroutine like this is a good way to hit an NPE when the system is shutting down or something else you didn't expect, this Goroutine is still running, and the logger is no longer initialized enough to work correctly. I'll show you a better way in a moment with the refactor I'm providing.

**Review by akutz — CHANGES_REQUESTED** [review]
Hi @dilyar85 ,

Thank you for this. However, please refactor your cache to use this generic implementation I whipped up -- https://github.com/akutz/vm-operator/blob/feature/generic-cache/pkg/util/cache.go. I'm not gonig to submit this as a PR, you can just take it for your PR. Please note there's a series of simple tests too, but please feel free to expand on them -- https://github.com/akutz/vm-operator/blob/feature/generic-cache/pkg/util/cache_test.go.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (akutz) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:1]: See this comment -- https://github.com/vmware-tanzu/vm-operator/pull/47#pullrequestreview-1277855425

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (bryanv) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go]: I wouldn't expect that to leave sufficient for the pod to not OOM.

It does seem that eventually that a combination of large OVF image(s) and/or mapped to many namespaces, we'll be forced to do MD-only watches and disable the ctrl-runtime client cache for NS-scoped image. That said, at present, if this cache doesn't contain cluster-scoped images, it is more or less another namespace worth of images.

**Comment by bryanv** [comment]
> > How effective is this cache on pod restart? Is this mostly effective when creating a new namespace with CL already for other namespaces? When there are more items than `maxSize` and during the re-sync, I'm not sure if updating `lastAccessTime` is really cache friend to how we'll access the cache.
> 
> @bryanv The purpose of implementing this cache is to avoid downloading duplicate library item when same CL is attached to multiple namespaces in a cluster. Agree with your suggestion on `lastAccessTime`. I have updated the `get` function to avoid changing the item time (so it's pure read-only) and only the `set` function will touch it now (when creating or updating the existing item).

In the steady state, won't this cache be empty the controllers will noop when the ContentVersion matches, and the cache 
expires items 30 minutes after they're touched, correct? For attachments to multiple namespaces, what order reconcile request ultimately end up in the controller's queue seems to dictate effectiveness: if the items from different namespaces are interweaved, an item in one namespace may not get a cache hit because it is still being downloaded for the same CL item in another namespace.

File: controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go
Comment (dilyar85) [reply controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go:263]: After a second thought, I think it might be better to make the cl/ccl-item itself as a param here so that we won't need to add more params or update the option struct if needed later.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:1]: I actually thought about making this cache generic or using a 3rd party cache library, however, I made it specific to the content-library as I didn't see any obvious usecase for storing other objects atm. I am taking a deeper look at your implementation above and will try adopting it in this PR. Thanks again for sharing it.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go:37]: Ah, I didn't realize that, thanks for pointing it out.

**Comment by dilyar85** [comment]
> In the steady state, won't this cache be empty the controllers will noop when the ContentVersion matches, and the cache expires items 30 minutes after they're touched, correct? For attachments to multiple namespaces, what order reconcile request ultimately end up in the controller's queue seems to dictate effectiveness: if the items from different namespaces are interweaved, an item in one namespace may not get a cache hit because it is still being downloaded for the same CL item in another namespace.

@bryanv Yes, you got it right on the steady state. The current implementation does need reconciliation requests to come in the right order for it to work effectively. But I believe we can improve this. Instead of having a lock on the whole cache, what about putting a lock on each cache item?

Here's how I am thinking it would work: let's say there are multiple namespaces tied to the same CL. This means multiple reconciliation requests will come in to sync the VMIs, some of which may involve downloading the same OVF file. If the first request results in a cache miss, that specific cache item will be locked while the OVF is being downloaded and saved in the cache. This way, other requests for that same OVF will be put on hold until the lock is released, avoiding cache misses during the OVF download. 

Let me know if this approach sounds viable to you, and if so I'll implement this change in Andrew's generic cache example above. Thanks.

File: controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go
Comment (akutz) [reply controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go:263]: It doesn't matter if you need to update the option struct. The point of the option struct is to make it so the function signature doesn't have to change and break things. Put the CL/CCL item into the option struct.

**Comment by akutz** [comment]
> > How effective is this cache on pod restart? Is this mostly effective when creating a new namespace with CL already for other namespaces? When there are more items than `maxSize` and during the re-sync, I'm not sure if updating `lastAccessTime` is really cache friend to how we'll access the cache.
> 
> @bryanv The purpose of implementing this cache is to avoid downloading duplicate library item when same CL is attached to multiple namespaces in a cluster. Agree with your suggestion on `lastAccessTime`. I have updated the `get` function to avoid changing the item time (so it's pure read-only) and only the `set` function will touch it now (when creating or updating the existing item).

There's a reason in the generic example that I changed `maxSize` to `maxItems`. The name `maxSize` is misleading and implies content size, not number of items.

**Comment by akutz** [comment]
> > In the steady state, won't this cache be empty the controllers will noop when the ContentVersion matches, and the cache expires items 30 minutes after they're touched, correct? For attachments to multiple namespaces, what order reconcile request ultimately end up in the controller's queue seems to dictate effectiveness: if the items from different namespaces are interweaved, an item in one namespace may not get a cache hit because it is still being downloaded for the same CL item in another namespace.
> 
> @bryanv Yes, you got it right on the steady state. The current implementation does need reconciliation requests to come in the right order for it to work effectively. But I believe we can improve this. Instead of having a lock on the whole cache, what about putting a lock on each cache item?
> 
> Here's how I am thinking it would work: let's say there are multiple namespaces tied to the same CL. This means multiple reconciliation requests will come in to sync the VMIs, some of which may involve downloading the same OVF file. If the first request results in a cache miss, that specific cache item will be locked while the OVF is being downloaded and saved in the cache. This way, other requests for that same OVF will be put on hold until the lock is released, avoiding cache misses during the OVF download.
> 
> Let me know if this approach sounds viable to you, and if so I'll implement this change in Andrew's generic cache example above. Thanks.

So what you need is a `sync.Map` that follows the pattern of `LoadOrStore`. For example, please see what I just pushed to:

* https://github.com/akutz/vm-operator/blob/feature/generic-cache/pkg/util/lock_pool.go
* https://github.com/akutz/vm-operator/blob/feature/generic-cache/pkg/util/lock_pool_test.go

These make it super easy to create and manage the locks necessary. You could easily add a TTL to them as well if you wanted.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_provider.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_provider.go]: @ericvmw I looked into it but the implementation would require a wider test framework change in the content library provider code. As this patch is focusing on the image cache I would address it hopefully in a future PR.

File: controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go
Comment (dilyar85) [reply controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go]: I ended up having just clitem and vmi (both in `ctrlruntime.Object` type) for this method's params. I don't expect it would change again given that all the sync info would be included in these objects.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go]: Updated this PR with your generic type cache included.

File: pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go
Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go]: Resolving this conversation as the `content_lirbary_cache` is replaced with the generic type cache.

File: pkg/util/lock_pool.go
Comment (bryanv) [inline pkg/util/lock_pool.go:45]: Maybe I'm missing it but is this being called anywhere?

File: pkg/util/cache.go
Comment (bryanv) [inline pkg/util/cache.go:95]: This is only called in the tests, right? Does that mean that the send done as part of the expire is going to block because there is nothing reading from the channel?

**Comment by bryanv** [comment]
In an earlier comment, I asked if this made sense for cluster scoped VMI but I didn't see that addressed or answered. Looks like this still caches all images. Sorry if I missed the response in the comments.

I'm still a little concerned about effectiveness and the visible behavior around this, ie, from a CU perspective, why is the behavior - when looking at the VC UI - different when creating a second NS 15 vs 45 min after the first NS? In most env, NS churn is not likely terribly common but I think that means the cache will be empty in the common case anyways b/c of the 30 min expiration?

IDK what the lift is for CL to provide this but I infer that they're cracking & caching part of the OVF based what's shown in the UI.

Don't treat any of my comments necessarily as blockers but I hope there is a longer term CL solution planned. Maybe an intermediate solution, implemented in VMOP, would be to set aside a separate vmware-system- NS, and use k8s as a durable cache, copying the VMIs from there (tracking CL item deletion vs CL un-association vs CL deletion would be at least a moderate PITA).

File: pkg/util/lock_pool.go
Comment (ericvmw) [inline pkg/util/lock_pool.go:13]: Does this pool need to be maintained outside of cache? the key is the same, and it looks the lifecycle of each lock pool entry needs to be in sync with the cache item, i.e. when item is cleaned, it needs to be removed from the map here.

File: pkg/util/lock_pool.go
Comment (dilyar85) [reply pkg/util/lock_pool.go:45]: I pushed a new commit that will delete the specific OVF item lock when it's expired in the cache, so this function is being used now (from `InitOvfCacheAndLockPool` within vmprovider.go file).

File: pkg/util/cache.go
Comment (dilyar85) [reply pkg/util/cache.go:95]: I don't think it's blocking anything as the expire send is always done in a goroutine. With that being said, the newly added `InitOvfCacheAndLockPool` will now read from this channel to clean up the corresponding lock item.

File: pkg/util/cache.go
Comment (dilyar85) [reply pkg/util/cache.go:29]: I pushed a new commit that will read from the expiredChan now.

File: pkg/util/cache.go
Comment (dilyar85) [reply pkg/util/cache.go:134]: We are adding this cache as a generic type under the `pkg/util`, not specifically to the OVF content. In our current use case, we check the content version but it could be possible to have a simple string id for other types of cache implementation in the future.

File: pkg/util/lock_pool.go
Comment (dilyar85) [reply pkg/util/lock_pool.go:13]: Good point. I have added a new function `InitOvfCacheAndLockPool` under the `vmprovider.go` file to clean up the lock when the same item expires from the cache.

File: pkg/util/lock_pool.go
Comment (dilyar85) [reply pkg/util/lock_pool.go:14]: I believe it's named as this way so that we can call `p.ool` in its methods. I will let @akutz confirm this as the code is's originally from his branch at https://github.com/akutz/vm-operator/blob/feature/generic-cache/pkg/util/lock_pool.go

Comment (ericvmw) [reply pkg/util/lock_pool.go:14]: I see, it looks good for private variables.

File: pkg/util/cache.go
Comment (ericvmw) [inline pkg/util/cache.go:36]: It would be helpful to document the cache behavior with regarding to the expireAfter and maxItems, also LastUpdated in CacheItem.

**Review by ericvmw — APPROVED** [review]

File: pkg/util/cache.go
Comment (dilyar85) [reply pkg/util/cache.go:36]: I think those are already documented in the `Cache` and `CacheItem` structs. Will keep the current comments for now as they are all in the same file.

**Review by akutz — APPROVED** [review]

File: pkg/util/cache.go
Comment (bryanv) [reply pkg/util/cache.go:95]: The goroutine will block because there is/was nothing reading from the channel.

File: pkg/vmprovider/providers/vsphere/vmprovider.go
Comment (bryanv) [inline pkg/vmprovider/providers/vsphere/vmprovider.go:101]: Isn't this racey: before the message is handled here, another call comes into getOvfEnvelope() and then an in-use lock is removed from this cache?

Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/vmprovider.go:101]: Yeah, but that shouldn't block us because the in-use lock can still be safely executed after removal, although it won't prevent another download of the same item. I considered locking it during deletion, but it wouldn't prevent the lock from being used after removal. Adding another layer of protection for the lock pool isn't ideal either, as there could be at most 2 duplicate downloads in the worst-case scenario. And this only occurs in rare cases when the lock is removed while in use.

Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/vmprovider.go:101]: On second thoughts, maybe locking and unlocking during deletion could still be useful. If getOvfEnvelope() can lock it, the lock won't be deleted until the item is downloaded and put into the cache, avoiding duplicate downloads. However, if delete() locks it first, there's still a small chance of duplicate downloads. I'll add comments in the updated code to document our discussion here.

Comment (bryanv) [reply pkg/vmprovider/providers/vsphere/vmprovider.go:100]: Yes, it will break the guarantee of no duplicate OVF retrieves we're trying to achieve in the first place by adding this outer locking pool.

Ultimately I think we're trying to make something more complicated that what's needed: the outer lock should be pushed into the CacheItem[T], and the Cache[T] has a Load/Fill function that populates the T, so we have something like https://guava.dev/releases/19.0/api/docs/com/google/common/cache/LoadingCache.html, and there's no need for the eviction channel to try to maintain consistency of the other cache.

Comment (dilyar85) [reply pkg/vmprovider/providers/vsphere/vmprovider.go:100]: I believe that's what I had in my original content library specific cache, where the item has its own lock... However, I understand this util cache doesn't have item-level locks as it may not be necessary for common use cases, where the cache is populated or retrieved quickly. It may not completely eliminate duplicate OVF retrieves, but it only occurs in rare edge cases where an OVF expires right after starting to download, and only two duplicates could happen even with multiple namespaces associated. If you feel strongly about this, I'm happy to discuss it further offline and make a follow-up change to address it.

**Pull request merged**
